FWCore/Services: do not optimize 'edmtest::FpeTester::analyze'

davidlt opened the pull request:
Resolves unit test (`fpe_test_2.sh`) failure in CLANG IB.

```
Completed cmsRun with DIVIDEBYZERO exception enabled
cmsRun status:  0
Test FAILED, status neither 136 nor 11
status = 256
```

`do_division` call is optimized out from `edmtest::FpeTester::analyze`.
It's caused by division by zero, which is undefined behavior thus
compiler is free to do what it wants, e.g., optimize it out.

Disable any optimizations on the function.

Signed-off-by: David Abdurachmanov David.Abdurachmanov@cern.ch

cmsbuild commented:
A new Pull Request was created by @davidlt  for CMSSW_7_5_X.

FWCore/Services: do not optimize 'edmtest::FpeTester::analyze'

It involves the following packages:

FWCore/Services


@cmsbuild, @Dr15Jones, @ktf, @nclopezo can you please review it and eventually sign? Thanks.
@Martin-Grunewald, @wddgit, @wmtan this is something you requested to watch as well.
You can sign-off by replying to this message having '+1' in the first line of your reply.
You can reject by replying  to this message having '-1' in the first line of your reply.
If you are a L2 or a release manager you can ask for tests by saying 'please test' in the first line of a comment.
@nclopezo you are the release manager for this.
You can merge this pull request by typing 'merge' in the first line of your comment.

cmsbuild commented:
-1 
Tested at: 3b2d740c69716de6a59389a997c93ab7c3a49a8a
 you can see the results of the tests here: 
 https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-8892/4411/summary.html

davidlt commented:
This does not affect relvals and all failures are `Step0-DAS_ERROR`.

Dr15Jones commented on FWCore/Services/test/FpeTester.cc:
```diff
@@ -54,6 +54,7 @@ namespace edmtest {
   }
 
   void
+  __attribute__((optnone))
```
Could we use the C++11 syntax instead? ```[[optnone]]```

davidlt replied:
After looking carefully into Clang testsuite, I found `[[clang::optnone]]`, but I don't know how GCC (or other compilers) would handle that.

Commit:
FWCore/Services: do not optimize 'edmtest::FpeTester::analyze'

Resolves unit test (`fpe_test_2.sh`) failure in CLANG IB.

    Completed cmsRun with DIVIDEBYZERO exception enabled
    cmsRun status:  0
    Test FAILED, status neither 136 nor 11
    status = 256

`do_division` call is optimized out from `edmtest::FpeTester::analyze`.
It's caused by division by zero, which is undefined behavior thus
compiler is free to do what it wants, e.g., optimize it out.

Disable any optimizations on the function.

Signed-off-by: David Abdurachmanov <David.Abdurachmanov@cern.ch>

davidlt commented:
Updated. Seems to work with GCC, but Clang was picky about attribute location. Even ICC was happy!

Dr15Jones commented:
pleaset test

Dr15Jones commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_7_5_X IBs unless changes or unless it breaks tests. This pull request requires discussion in the ORP meeting before it's merged. @davidlange6, @nclopezo, @smuzaffar

cmsbuild commented:
+1
Tested at: 93ab028d321d68585deeaf06720d09b13db6a69f
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-8892/4467/summary.html

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_7_5_X IBs unless changes (tests are also fine). This pull request requires discussion in the ORP meeting before it's merged. @davidlange6, @nclopezo, @smuzaffar

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-8892/4467/summary.html

The workflows 140.53 have different files in step1_dasquery.log than the ones found in the baseline. You may want to check and retrigger the tests if necessary. You can check it in the "files" directory in the results of the comparisons

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_7_5_X IBs unless changes (tests are also fine). This pull request requires discussion in the ORP meeting before it's merged. @davidlange6, @nclopezo, @smuzaffar

davidlange6 commented:
+1
